### PR TITLE
[SWE-36] Feed Backend

### PR DESCRIPTION
**Pull request by datderek**
Task ID: [SWE-36](https://linear.app/swe-dem-boys/issue/SWE-36/feed-backend)
## PR Description
Implement the backend such that the frontend can request and displays posts from the user's following and user's own.

## Changes Implemented
- API Gateway/Lambda function set up to query RDS for posts
- Posts are returned from the API in chronological order
- Posts are requested on initial page load and scroll
- Postman API updated accordingly
## Notes
We will need additional Linear tasks for the following:
- Currently requesting posts from a static set of users as we do not have "following" relationships set-up in the GraphDB yet
- Code queries for post by username of the user that created it
  - Confused about how we implement `user_ids` into our webapp
  - i.e. Do we need to query for a user_ids (in the GraphDB) everytime we create a post in order to store it in the relational DB?
    - Atleast with usernames, they are stored in the Cognito identity token

### CODE REVIEW

**Review by sammyoto — APPROVED** [review]
Looks good to me. One minor issue, not very important but if you scroll to the very bottom if there are no more posts the loading wheel keeps loading forever. Might want to add something that says no more posts or something when loaded. Other than that looks good. Approved. Might also want an @jackfales approval on this one since it's big.

File: app/dashboard/page.tsx [hunk 1]
```diff
@@ -40,17 +30,28 @@ export default async function Dashboard() {
...
     <Navbar username={username}></Navbar>
     <main className='bg-cream-100 flex items-center justify-center pt-14'>
-      <div className='flex-[0_1_670px] flex flex-col items-center'>
+      <div className='flex-[0_1_670px] flex flex-col items-center pb-7'>
         <CreatePostModal/>
         {
-          fetchPosts(0, testData).map((post, index) => (
-            <Post name={post['name']} key={index}></Post>
+          posts.map((post, index) => (
+            <Post imageUrl={post[0]} title={post[1]} createdBy={post[6]} key={index}></Post>
```
Comment (jackfales) [inline app/dashboard/page.tsx:51]: Replace array references with constant values for readability.

Before:
`imageUrl={post[0]}`

After: 
`const IMAGE_URL_INDEX = 0;`
`imageUrl={post[IMAGE_URL_INDEX]}`

This helps with readability and easier to update in the event a post's structure changes.

File: app/dashboard/page.tsx [hunk 2]
```diff
@@ -40,17 +30,28 @@ export default async function Dashboard() {
     redirect('/login');
   }
   
+  // TODO(SWE-67): Grab posts from following users
+  // Sends a request to load the initial posts
+  const payload = { "userIds": ['dtran', 'jfales', 'sfales'] };
```
Comment (jackfales) [inline app/dashboard/page.tsx:35]: I thought userIds were supposed to be either (1) hash values or (2) graph node ids? I understand the current values are hardcoded, but aren't these example userIds actually usernames?

File: app/components/LoadMore.tsx [hunk 3]
```diff
@@ -4,26 +4,34 @@ import { useEffect, useState } from "react";
...
+export default function LoadMore() {
   const [posts, setPosts] = useState<Object[]>([]);
   const [pagesLoaded, setPagesLoaded] = useState(0);
+  const [hasMorePosts, setHasMorePosts] = useState(true);
 
-  // TODO(SWE-36): Call to Route handler instead of `fetchPosts`
-  /**
+  /*
    * Grabs posts associated with the next page then updates the `pagesLoaded`
    * and `posts` state
```
Comment (jackfales) [inline app/components/LoadMore.tsx:15]: What does it update the states with?

File: app/components/LoadMore.tsx [hunk 4]
```diff
@@ -38,13 +46,11 @@ export default function LoadMore({postsData}: {postsData: Object[]}) {
   }, [inView]);
 
   return (<>
-    {
-      posts.map((post, index) => (
-        <Post name={post['name']} key={index}></Post>
-      ))
-    }
+    {posts.map((post, index) => (
+        <Post imageUrl={post[0]} title={post[1]} createdBy={post[6]} key={index}></Post>
```
Comment (jackfales) [inline app/components/LoadMore.tsx:50]: Same as above

File: app/components/LoadMore.tsx [hunk 5]
```diff
@@ -4,26 +4,34 @@ import { useEffect, useState } from "react";
...
+  const [hasMorePosts, setHasMorePosts] = useState(true);
 
-  // TODO(SWE-36): Call to Route handler instead of `fetchPosts`
-  /**
+  /*
    * Grabs posts associated with the next page then updates the `pagesLoaded`
    * and `posts` state
    */ 
-  const loadMore = () => {
+  const loadMore = async () => {
```
Comment (jackfales) [inline app/components/LoadMore.tsx:17]:
```suggestion
  const loadMorePosts = async () => {
```

**Review by jackfales — COMMENTED** [review]
Couple of questions, PTAL

File: app/dashboard/page.tsx (hunk 2 above)
Comment (datderek) [reply app/dashboard/page.tsx:35]: These will be implemented in @sammyoto's PR `[SWE-63]`

File: app/api/posts/users/route.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+ */
+export async function POST(request: NextRequest) {
+  const searchParams: URLSearchParams = request.nextUrl.searchParams;
+  const page: String | null = searchParams.get('page');
+  const pageSize: String | null = searchParams.get('pageSize');
+  const body = await request.json();
+
+  // Forwards the request to the API Gateway which should return the post information for all the users
+  const res = await fetch(`https://wb07xao9oa.execute-api.us-west-2.amazonaws.com/dev/posts/{users+}?page=${page}&pageSize=${pageSize}`, {
+    method: 'POST',
```
[the remark this answers is not in the document]
Comment (datderek) [reply app/api/posts/users/route.ts:14]: Changed method to `GET` as it aligns better with what is expected of the request nature. Updated API gateway accordingly.

[the remark this answers is not in the document]
Comment (datderek) [reply app/api/posts/users/route.ts:14]: Did some more research, and I think I will be keeping it as `POST`. 

This [article](https://evertpot.com/get-request-bodies/) discusses how `GET` request bodies are undefined behavior and how they can potentially be dropped.

Also I tried to switch the request type in API Gateway and it looks like it doesn't support a request body when switching from `POST` to `GET` - atleast in terms of sending a test request. There is also this [documentation](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/RequestAndResponseBehaviorCustomOrigin.html#RequestCustom-get-body) that shows how Amazon Cloud Front (I couldn't find an API Gateway specific one), denies `GET` request with bodies.

**Comment by datderek** [comment]
Left the request as `POST` as some further research supports against using `GET`. 

This [article](https://www.ietf.org/archive/id/draft-ietf-httpbis-safe-method-w-body-03.txt) by the IETF highlights the use of a new HTTP method `QUERY` that addresses this exact problem (sending a `GET` with a request body). However, this method does not seem like it is supported by API Gateway or Postman yet. Therefore, I decided to just continue using `POST`.

File: amplify/backend/function/cognitoCreateNetpuneUser/src/index.py
Comment (sammyoto) [inline amplify/backend/function/cognitoCreateNetpuneUser/src/index.py:1]: I thought we were just going to use postUser? If we are using postUser we don't need to add another function. You can add postUser as a pre sign up function on cognito.

**Review by sammyoto — CHANGES_REQUESTED** [review]
Overall I agree with your decision to use POST instead of GET. One thing, we don't need to add cognitoCreateNeptuneUser, we already have postUser.

File: amplify/backend/function/cognitoCreateNetpuneUser/src/index.py
Comment (datderek) [reply amplify/backend/function/cognitoCreateNetpuneUser/src/index.py:1]: I agree, I don't believe I created `cognitoCreateNeptuneUser`. Want me to delete it from the Amplify build though?

**Review by sammyoto — APPROVED** [review]
Clean. Good work

**Pull request merged**